eyetracking/bids_scripts/friends_driftCor.py

```diff
@@ -0,0 +1,551 @@
+    '''
+    frame_nums = []
+
+    et_x = []
+    et_y = []
+
+    dist_x = []
+    dist_y = []
+
+    dg_x = []
+    dg_y = []
+
+    for i in range(frame_count):
+        if len(gazes_perframe[i]) > 0:
+            x_et, y_et = np.mean(np.array(gazes_perframe[i])[:, 1:], axis=0)
+
+            if use_deepgaze:
+                """
+                ONLY use only "high confidence" Deepgaze points for frames in
+                which a single local maximum was identified
+                (>.8 max, >20 pixels of distance).
+                """
+                if len(DGvals_perframe[i]) == 1:
+                    x_dg, y_dg = DGvals_perframe[i][0][1:]
+
+                    dist_x.append(x_et - x_dg)
+                    dist_y.append(y_et - y_dg)
+
+                    et_x.append(x_et)
+                    et_y.append(y_et)
+
+                    dg_x.append(x_dg)
+                    dg_y.append(y_dg)
+
+                    frame_nums.append(i)
+
+            else:
+                et_x.append(x_et)
+                et_y.append(y_et)
+
+                dist_x.append(x_et - x_mass)
+                dist_y.append(y_et - y_mass)
+
+                frame_nums.append(i)
+
+    return (et_x, et_y), (dist_x, dist_y), (dg_x, dg_y), frame_nums
+
+
+def median_clean(frame_times, et_x, et_y, dist_x, dist_y):
+    '''
+    Within bins of 1/100 the number of frames,
+    select only frames where distance between gaze and deepgaze falls within 0.6 stdev of the median
+    These frames most likely reflect when deepgaze and gaze "look" at the same thing
+    '''
+    jump = int(len(dist_x)/100)
+    idx = 0
+    gap = 0.6 # interval of distances included around median, in stdev
+
+    filtered_times = []
+    filtered_x = []
+    filtered_y = []
+    filtered_distx = []
+    filtered_disty = []
+
+    current_medx = np.median(np.array(dist_x)[idx:idx+jump])
+    current_medy = np.median(np.array(dist_y)[idx:idx+jump])
+    stdevx = np.std(np.array(dist_x)[idx:idx+jump])
+    stdevy = np.std(np.array(dist_y)[idx:idx+jump])
+
+    for i in range(len(dist_x)):
+        if i > (idx + jump):
+            idx += 1
+            current_medx = np.median(np.array(dist_x)[idx:idx+jump])
+            current_medy = np.median(np.array(dist_y)[idx:idx+jump])
+            stdevx = np.std(np.array(dist_x)[idx:idx+jump])
+            stdevy = np.std(np.array(dist_y)[idx:idx+jump])
+
+        if dist_x[i] < current_medx + (gap*stdevx):
+            if dist_x[i] > current_medx - (gap*stdevx):
+                if dist_y[i] < current_medy + (gap*stdevy):
+                    if dist_y[i] > current_medy - (gap*stdevx):
+                        filtered_times.append(frame_times[i])
+                        filtered_distx.append(dist_x[i])
+                        filtered_disty.append(dist_y[i])
+                        filtered_x.append(et_x[i])
+                        filtered_y.append(et_y[i])
+
+    return filtered_times, filtered_x, filtered_y, filtered_distx, filtered_disty
```

Comment:
wowowow that's a scary function.
- The median estimation window seems not centered as it is `[i:i+100]` which I now remember having detected from the graphs then forgot to ask you. 
- why not using `scipy.signal.medfilt` ?
- seems like you take 100 frames whatever their time, so they will be border effects with missing data (camera freeze).
- why the double indexing (`i`/`idx`) ? 


Maybe something like the following would do the same but using numpy array (way faster) and without the windowing issue.
(I have not tested the code, but that would be the way I interpret the code above)

```suggestion
from scipy.signal import medfilt
from scipy.ndimage.filters  import uniform_filter

def window_stdev(arr, radius):
    c1 = uniform_filter(arr, radius*2, mode='constant', origin=-radius)
    c2 = uniform_filter(arr*arr, radius*2, mode='constant', origin=-radius)
    return ((c2 - c1*c1)**.5)[:-radius*2+1,:-radius*2+1]

def median_clean(frame_times, et_x, et_y, dist_x, dist_y, gap=0.6):
    dist_x_med = medfilt(dist_x, 100)
    dist_y_med = medfilt(dist_y, 100)
    
    dist_x_std = window_stdev(dist_x, 50)
    dist_y_std = window_stdev(dist_y, 50)
    
    filter_mask = np.logical_and(
        np.abs(dist_x_med-dist_x_std) < gap*dist_x_std,
        np.abs(dist_y_med-dist_y_std) < gap*dist_y_std
        )
        
    return frame_times[filter_mask], et_x[filter_mask], et_y[filter_mask], dist_x[filter_mask], dist_y[filter_mask]
```

Reply:
argh, it's supposed to be idx += jump, not idx += 1... (that is a very crude function to begin with, I meant to speed it up by not recalculating the median at every point. it's shocking that it worked at all! I meant for the window to be centered on average but this obviously isn't it. And yes, it would be better to consider the timestamps to handle camera freezes. Thanks for picking up all of this, I've meant to improve on this monstruosity for a while and just got lazy).